tajo-core/src/main/java/org/apache/tajo/engine/planner/physical/WindowAggExec.java

```diff
@@ -347,6 +350,372 @@ private void finalizeWindow() {
+    int numberOfTuples = 0;
+
+    Tuple inTuple = accumulatedInTuples.get(startOffset);
+    if (isForward) {
+      do {
+        numberOfTuples++;
+        startOffset++;
+      } while (startOffset < accumulatedInTuples.size() && comp.compare(accumulatedInTuples.get(startOffset), inTuple) == 0);
+    } else {      // backward direction
+      do {
+        numberOfTuples++;
+        startOffset--;
+      } while (startOffset >= 0 && comp.compare(accumulatedInTuples.get(startOffset), inTuple) == 0);
+    }
+
+    return numberOfTuples;
+  }
+
+  private Tuple getFunctionSpecificInput(String funcName, int dataStart, int dataEnd) {
+    if (funcName.equals("first_value")) {
+      // check the frame start is within the partition
+      if (dataStart <= dataEnd && dataStart >= 0 && dataStart < accumulatedInTuples.size()) {
+        return accumulatedInTuples.get(dataStart);
+      }
+    } else if (funcName.equals("last_value")) {
+      // check the frame end is within the partition
+      if (dataStart <= dataEnd && dataEnd >= 0 && dataEnd < accumulatedInTuples.size()) {
+        return accumulatedInTuples.get(dataEnd);
+      }
+    }
+    return null;
+  }
+
+  private void evaluateAggregationFunction(int functionIndex, LogicalWindowSpec.LogicalWindowFrame windowFrame,
+                                           BaseTupleComparator comp) {
+    LogicalWindowSpec.LogicalWindowFrame.WindowFrameType windowFrameType = windowFrame.getFrameType();
+    WindowSpec.WindowFrameUnit windowFrameUnit = windowFrame.getFrameUnit();
+    int windowFrameStartOffset = windowFrame.getStartBound().getNumber();
+    int windowFrameEndOffset = windowFrame.getEndBound().getNumber();
+    int frameStart, frameEnd;
+
+    switch(windowFrameType) {
+      case ENTIRE_PARTITION:
+      {
+        Datum result = rangedAggregation(functionIndex, 0, accumulatedInTuples.size());
+
+        for (int j = 0; j < evaluatedTuples.size(); j++) {
+          Tuple outTuple = evaluatedTuples.get(j);
+
+          outTuple.put(nonFunctionColumnNum + functionIndex, result);
+        }
+        break;
+      }
+      case TO_CURRENT_ROW:
+      {
+        if (windowFrameUnit == WindowSpec.WindowFrameUnit.RANGE) {
+          // If RANGE is used, it is guaranteed that windowFrameEndOffset is 0
+          int sameEndRange = 0;
+          for (int i = 0; i < accumulatedInTuples.size(); i++) {
+            // including all rows that has the same order by value
+            if (sameEndRange == 0) {
+              sameEndRange = aggregatingTuplesWithSameKey(functionIndex, i, comp, true);
+            }
+            sameEndRange --;
+
+            Datum result = functions[functionIndex].terminate(contexts[functionIndex]);
+            Tuple outTuple = evaluatedTuples.get(i);
+            outTuple.put(nonFunctionColumnNum + functionIndex, result);
+          }
+        } else {
+          if (windowFrameEndOffset > 0) {
+            int i =0; int j = 0;
```

Comment:
Would you mind explaning on this block? 
In addition, it would be better to rename ```i```, and ```j``` to be more intuitive if possible.